tests/test_builtin_traits.py

```diff
@@ -113,40 +124,133 @@ class HasNoTerminatorOp(IRDLOperation):
     """
-    Test that an operation with a NoTerminator trait expects its containing
-    region to have an empty single block and no terminator.
+    Test that an operation with an IsTerminator trait may not have successor
+    blocks.
     """
+    block0 = Block([])
+    block1 = Block([IsTerminatorOp.create()])
+    region0 = Region([block0, block1])
+    op0 = TestOp.create(regions=[region0])
 
-    noterm_op = HasNoTerminatorOp.build(regions=[[Block()]])
-    noterm_op.verify()
+    op0.verify()
 
 
-def test_has_no_terminator_verify_with_non_empty_single_block_region():
+def test_is_terminator_fails_if_not_last_operation_parent_block():
     """
-    Test that an operation with a NoTerminator trait expects its containing
-    region to have a non-empty single block and no terminator.
+    Test that an operation with an IsTerminator trait fails if it is not the
+    last operation in its parent block.
     """
+    block0 = Block([IsTerminatorOp.create(), TestOp.create()])
+    region0 = Region([block0])
+    op0 = TestOp.create(regions=[region0])
 
-    noterm_op = HasNoTerminatorOp.build(regions=[[Block([TestOp.create()])]])
-    noterm_op.verify()
+    with pytest.raises(
+        VerifyException, match="must be the last operation in the parent block"
+    ):
+        op0.verify()
 
 
-def test_has_no_terminator_not_single_block_region():
+@irdl_op_definition
+class IsolatedFromAboveOp(IRDLOperation):
     """
-    Test that an operation with a NoTerminator trait fails when its containing
-    region has more than single block.
+    An isolated from above operation.
     """
 
-    block0 = Block([TestOp.create(), TestOp.create()])
-    block1 = Block([TestOp.create(), TestOp.create()])
-    noterm_op = HasNoTerminatorOp.build(regions=[[block0, block1]])
+    name = "test.isolated_from_above"
+
+    region: Region = region_def()
+
+    traits = frozenset([IsolatedFromAbove()])
+
 
-    with pytest.raises(VerifyException, match="expects single block region"):
-        noterm_op.verify()
+def test_isolated_from_above():
+    # Empty Isolated is fine
+    op = IsolatedFromAboveOp(regions=[Region()])
+    op.verify()
+
+    block = Block(arg_types=[builtin.i32])
+    block.add_op(arith.Addi(block.args[0], block.args[0]))
+
+    # Test a simple, properly Isolated
+    op = IsolatedFromAboveOp(regions=[Region([block])])
+    op.verify
+
+    # Check a simple isolation violation
+    out_cst = arith.Constant.from_int_and_width(0, builtin.i32)
+    out_block = Block(
+        [
+            out_cst,
+            IsolatedFromAboveOp(
+                regions=[Region(Block([arith.Addi(out_cst, out_cst)]))]
+            ),
+        ]
+    )
+    message = "Operation using value defined out of its IsolatedFromAbove parent!"
+    with pytest.raises(VerifyException, match=message):
+        out_block.verify()
+
+    # Check a nested isolation violation
+    out_cst = arith.Constant.from_int_and_width(0, builtin.i32)
+    out_block = Block(
```

Comment:
I'd recommend implicit builders here to make it more readable

Reply:
This test wasn't added by this PR and I'd like to keep changes to a minimal set, especially when they touch so many other places.
A PR on improving test readability could be separate IMHO.